operator/pkg/ingress/ingress.go

```diff
@@ -0,0 +1,447 @@
+	}
+	return true
+}
+
+func (ic *IngressController) handleIngressAddedEvent(event ingressAddedEvent) error {
+	ingressClass := event.ingress.Spec.IngressClassName
+	if ingressClass == nil || *ingressClass != ciliumIngressClassName {
+		return nil
+	}
+	if err := ic.createEnvoyConfig(event.ingress); err != nil {
+		log.WithError(err).Warn("failed to create CiliumEnvoyConfig")
+		return err
+	}
+	if err := ic.createEndpoints(event.ingress); err != nil {
+		log.WithError(err).Warn("failed to create endpoints")
+		return err
+	}
+	if err := ic.createLoadBalancer(event.ingress); err != nil {
+		log.WithError(err).Warn("failed to create load balancer")
+		return err
+	}
+	return nil
+}
+
+func (ic *IngressController) handleIngressUpdatedEvent(event ingressUpdatedEvent) error {
+	ingressClass := event.newIngress.Spec.IngressClassName
+	if ingressClass == nil || *ingressClass != ciliumIngressClassName {
+		return nil
+	}
+	if err := ic.createEnvoyConfig(event.newIngress); err != nil {
+		log.WithError(err).Warn("failed to update CiliumEnvoyConfig")
+		return err
+	}
+	if err := ic.createEndpoints(event.newIngress); err != nil {
+		log.WithError(err).Warn("failed to update endpoints")
+		return err
+	}
+	if err := ic.createLoadBalancer(event.newIngress); err != nil {
+		log.WithError(err).Warn("failed to update load balancer")
+		return err
+	}
+	return nil
+}
+
+func (ic *IngressController) handleIngressDeletedEvent(event ingressDeletedEvent) error {
+	// as ciliumEnvoyConfig is non-namespace scoped, it should be removed manually.
+	// other resources (e.g service, endpoints) will be removed automatically via ownerReferences
+	log.WithField("ingress", event.ingress.Name).Info("Deleting CiliumEnvoyConfig for ingress")
+	if err := ic.deleteCiliumEnvoyConfig(event.ingress); err != nil {
+		log.WithError(err).Warn("failed to delete cilium-envoy-config")
+		return err
+	}
+	return nil
+}
+
+func (ic *IngressController) handleIngressServiceUpdatedEvent(ingressServiceUpdated ingressServiceUpdatedEvent) error {
+	service := ingressServiceUpdated.ingressService
+	serviceKey, err := cache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		return err
+	}
+	ingressKey := getIngressKeyForService(service)
+	ingress, err := ic.getByKey(ingressKey)
+	if err != nil {
+		log.WithError(err).WithField("service", serviceKey).Warn("Failed to lookup ingress for service")
+		return err
+	}
+	scopedLog := log.WithField("ingress", ingressKey).WithField("service", serviceKey).
+		WithField("svc-status", service.Status.LoadBalancer).
+		WithField("ingress-status", ingress.Status.LoadBalancer)
+
+	if reflect.DeepEqual(service.Status.LoadBalancer, ingress.Status.LoadBalancer) {
+		scopedLog.Info("No change in load balancer status")
+		return nil
+	}
+
+	newIngressStatus := getIngressForStatusUpdate(ingress, service.Status.LoadBalancer)
+	_, err = k8s.Client().NetworkingV1().Ingresses(ingress.Namespace).UpdateStatus(context.Background(), newIngressStatus, metav1.UpdateOptions{})
```

Comment:
Not blocking but would be great to know if we could use `ApplyStatus` instead perhaps it would even prevent us from having to generate an ingress object with `getIngressForStatusUpdate`? It looks like `ApplyStatus` does not need a full Ingress object, only its name and status.

Reply:
@sayboras Could you respond to Andre question?

Reply:
oh I missed this, let me track this work along with standard informer pattern as per snippet shared by Andre before.